### PR TITLE
feat(repository-tests): run repository-tests on cloudant

### PR DESCRIPTION
**Pull request by agnes512**
closes https://github.com/strongloop/loopback-next/issues/3437

Create `acceptance-repository-cloudant` folder.
- Modify tests in `replace-by-id` need to be modified to run the replacement operation on Cloudant ( it needs an extra property `_rev` to do so).
- Set up Travis task: it needs to run the script(`docker.setup.js`) to set up the docker image.
- the `docker.setup.js` file is based on @jannyHou 's [PR](https://github.com/strongloop/loopback-connector-cloudant/pull/227#pullrequestreview-317888999). I modified it a bit, now it can be used to set up the local test as well with customized names. ( see README).
- Add local test setup and README file to allow run repo tests on Cloudant ( or on docker) locally.

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,9 +1,9 @@
 sudo: false
 language: node_js
 node_js:
-  - "8"
-  - "10"
-  - "12"
+  - '8'
+  - '10'
+  - '12'
```
Comment (bajtos) [inline .travis.yml]: Please revert unrelated changes, or alternatively open a standalone pull request to make them.

What VSCode plugin do you use to format the YAML file? Maybe we need to tell Prettier to ignore `.travis.yml`?

File: acceptance/repository-cloudant/src/__tests__/cloudant-default-repository.acceptance.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+// License text available at https://opensource.org/licenses/MIT
+
+import {DefaultTransactionalRepository} from '@loopback/repository';
+import {
+  CrudRepositoryCtor,
+  crudRepositoryTestSuite,
+} from '@loopback/repository-tests';
+import {CLOUDANT_CONFIG, CLOUDANT_FEATURES} from './cloudant.datasource';
+
+describe('CLOUDANT + DefaultTransactionalRepository', () => {
```
Comment (bajtos) [inline acceptance/repository-cloudant/src/__tests__/cloudant-default-repository.acceptance.ts]: Since Cloudant does not support transactions, let's use `DefaultCrudRepository` as we already do in MongoDB tests.

https://github.com/strongloop/loopback-next/blob/9def622d7bafb7eb2470cd896372f030998e3fd4/acceptance/repository-mongodb/src/__tests__/mongodb-default-repository.acceptance.ts#L6-L20

File: .travis.yml (hunk 1 above)
Comment (agnes512) [reply .travis.yml]: Yes, I use Prettier

File: packages/repository-tests/src/crud/replace-by-id.suite.ts [hunk 3]
```diff
@@ -34,6 +34,11 @@ export function createSuiteForReplaceById(
     })
     id: MixedIdType;
 
+    // cloudant needs this property to do replacement method
+    // see cloundant README file for more details
```
Comment (agnes512) [inline packages/repository-tests/src/crud/replace-by-id.suite.ts]: [Cloudant README](https://github.com/strongloop/loopback-connector-cloudant#_rev-property)

File: .travis.yml (hunk 1 above)
Comment (agnes512) [reply .travis.yml]: ps: `.prettierignore` doesn't work.

File: acceptance/repository-cloudant/package.json [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    "verify": "npm pack && tar xf loopback-test-repository-cloudant*.tgz && tree package && npm run clean"
+  },
+  "author": "IBM Corp.",
+  "copyright.owner": "IBM Corp.",
+  "license": "MIT",
+  "devDependencies": {
+    "@loopback/build": "^1.7.1",
+    "@loopback/eslint-config": "^4.1.2",
+    "@loopback/repository": "^1.15.2",
+    "@loopback/repository-tests": "^0.5.2",
```
Comment (rmg) [inline acceptance/repository-cloudant/package.json]: The version inside this repo is currently `0.6.1` which is not matched by `^0.5.2` so when `npm run build -- --scope "@loopback/test-repository-cloudant" --include-dependencies` is run it doesn't detect that it should be building `packages/test-repositories` and then when it goes to run the tests using that package, it has no `./dist` directory.

File: .travis.yml [hunk 5]
```diff
@@ -80,6 +80,20 @@ matrix:
         - npm run build -- --scope "@loopback/test-repository-postgresql" --include-dependencies
         - psql -U postgres -a -f acceptance/repository-postgresql/test/schema.sql
         - cd acceptance/repository-postgresql && npm run mocha
+    - node_js: '10'
+      os: linux
+      env:
+        - TASK=test-repository-cloudant
+      install:
+        # - rm -rf node_modules && yarn cache clean
+        - npm ci && npm run docker:setup && source cloudant-config.sh && pwd
```
Comment (rmg) [inline .travis.yml]: This should be done as the first `script` step, I think.

File: .travis.yml [hunk 6]
```diff
@@ -80,6 +80,20 @@ matrix:
...
+
+      script:
+        - source cloudant-config.sh
```
Comment (rmg) [inline .travis.yml]: This line alone does nothing as far as I can tell since each `script` command should be run in a new shell so the env vars exported by `cloudant-config.sh` just go away after this line.

File: .travis.yml [hunk 7]
```diff
@@ -80,6 +80,20 @@ matrix:
...
+        - npm run postinstall -- --scope "@loopback/test-repository-cloudant" --include-dependencies
+        - npm run postinstall -- --include-dependencies
```
Comment (rmg) [inline .travis.yml]: Not necessary once dependencies are fixed

File: .travis.yml [hunk 8]
```diff
@@ -80,6 +80,20 @@ matrix:
...
+        - npm run build #-- --scope "@loopback/test-repository-cloudant" --include-dependencies
```
Comment (rmg) [inline .travis.yml]: `npm run build -- --scope "@loopback/test-repository-cloudant" --include-dependencies` should be sufficient here

File: .travis.yml [hunk 9]
```diff
@@ -80,6 +80,20 @@ matrix:
...
+        - cd acceptance/repository-cloudant && npm run mocha
```
Comment (rmg) [inline .travis.yml]: This is where you should source your config.. `source cloudant-config.sh && cd acceptance/repository-cloudant && npm run mocha`

File: acceptance/repository-cloudant/package.json (hunk 4 above)
Comment (agnes512) [reply acceptance/repository-cloudant/package.json]: Yes! it works once I updated it. Thank you so much!!

File: .travis.yml (hunk 8 above)
Comment (agnes512) [reply .travis.yml]: You're right. And after removing it, it takes less time to run the tests.

File: .travis.yml [hunk 10]
```diff
@@ -80,6 +80,16 @@ matrix:
         - npm run build -- --scope "@loopback/test-repository-postgresql" --include-dependencies
         - psql -U postgres -a -f acceptance/repository-postgresql/test/schema.sql
         - cd acceptance/repository-postgresql && npm run mocha
+    - node_js: "10"
+      os: linux
+      env:
+        - TASK=test-repository-cloudant
+      install:
```
Comment (rmg) [inline .travis.yml]: No need for any `install` steps here at all. Just need to add `npm run docker:setup` to `scripts` any time before the last step.

File: package.json [hunk 11]
```diff
@@ -62,7 +66,8 @@
     "docs:prepare": "./docs/bin/build-preview-site.sh",
     "docs:start": "cd docs/_preview && bundle exec jekyll serve --no-w --i",
     "mocha": "node packages/build/bin/run-mocha \"packages/*/dist/__tests__/**/*.js\" \"extensions/*/dist/__tests__/**/*.js\" \"examples/*/dist/__tests__/**/*.js\" \"packages/cli/test/**/*.js\" \"packages/build/test/*/*.js\"",
-    "posttest": "npm run lint"
+    "posttest": "npm run lint",
+    "docker:setup": "node ./docker.setup.js"
```
Comment (bajtos) [inline package.json]: Since the Docker setup is specific to Cloudant, can we move this npm script to `acceptance/repository-cloudant/package.json` please?

File: packages/repository-tests/src/crud/replace-by-id.suite.ts (hunk 3 above)
Comment (bajtos) [reply packages/repository-tests/src/crud/replace-by-id.suite.ts]: I find it suboptimal to force all connector tests (SQL, MongoDB, etc.) to use `_rev` property. Let's find a way how to allow individual connectors to provide any extra properties via `CrudFeatures`.

File: packages/repository-tests/src/crud/replace-by-id.suite.ts [hunk 12]
```diff
@@ -71,7 +76,9 @@ export function createSuiteForReplaceById(
       await repo.replaceById(created.id, created);
 
       const found = await repo.findById(created.id);
-      expect(toJSON(found)).to.deepEqual(
+      // use containDeep instead of deepEqual because cloudant changes _rev value after
+      // replacement
+      expect(toJSON(found)).to.containDeep(
```
Comment (bajtos) [inline packages/repository-tests/src/crud/replace-by-id.suite.ts]: I am concerned that by changing from `deepEqual` to `containDeep` we are allowing many more changes than just a different `_rev` value. For example, if the connector and/or the database adds new properties, this assertion will happily accept that, even though such behavior is most likely not wanted.

Let's find a way how to deal with `_rev` changes via a new `CrudFeatures` flag in such way that only changes in the `_rev` property is allowed and only when testing against Cloudant/CouchDB.

File: packages/repository-tests/src/crud/replace-by-id.suite.ts [hunk 13]
```diff
@@ -100,7 +107,9 @@ export function createSuiteForReplaceById(
...
```
Comment (bajtos) [inline packages/repository-tests/src/crud/replace-by-id.suite.ts]: Ditto.

File: acceptance/repository-cloudant/README.md [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+NOTICE: we use `couchDB3` docker image for testing Cloudant because Cloudant
+doesn't have a maintained image, and most of the their functionalities are the
+same (except couchDB doesn't support geosearch).
+
+- Install [Docker](https://docs.docker.com/engine/installation/).
+
+- Use the `couchDB3:latest` image, and rename the container as
+  `cloudant-testdb`.
+
+- Start the container, and create an admin and password for this image at
```
Comment (bajtos) [inline acceptance/repository-cloudant/README.md]: Can we automate this and the previous step please? In other repository acceptance packages, the docker-based setup requires the developer to run a single script only. Can we perhaps tell people to use `npm docker:setup`? (See also my other comments suggesting to move that script to this package.)

File: docker.setup.js [hunk 15]
```diff
@@ -0,0 +1,310 @@
...
+const docker = new require('dockerode')();
+const fmt = require('util').format;
+const http = require('http');
+const ms = require('ms');
+const fs = require('fs-extra');
+
+// // we don't pass any node flags, so we can call _mocha instead the wrapper
+// const mochaBin = require.resolve('mocha/bin/_mocha');
+
+process.env.CLOUDANT_DATABASE = 'testdb';
```
[the remark this answers is not in the document]
Comment (agnes512) [reply docker.setup.js]: This file will be removed once the release is done. The script is from https://github.com/strongloop/loopback-next/pull/3968. I use it here to check if it sets up travis correctly and takes reasonable time to finish.

File: docker.setup.js [hunk 16]
```diff
@@ -0,0 +1,310 @@
+// Copyright IBM Corp. 2017,2019. All Rights Reserved.
+// Node module: loopback-connector-cloudant
+// This file is licensed under the Artistic License 2.0.
+// License text available at https://opensource.org/licenses/Artistic-2.0
+
+'use strict';
+
+const _ = require('lodash');
+const async = require('async');
```
[the remark this answers is not in the document]
Comment (agnes512) [reply docker.setup.js]: We can make changes in that script if needed.

File: package.json (hunk 11 above)
Comment (agnes512) [reply package.json]: Yes :smile: I mentioned in the description above that once the script is released in `connector-cloudant`, the path will be changed to `"docker:setup": "node node_modules/loopback-connector-cloudant/docker.setup.js" `

File: packages/repository-tests/src/crud/replace-by-id.suite.ts (hunk 12 above)
Comment (agnes512) [reply packages/repository-tests/src/crud/replace-by-id.suite.ts]: Updated in the fifth commit.

File: acceptance/repository-cloudant/docker.setup.js [hunk 17]
```diff
@@ -0,0 +1,321 @@
...
+// License text available at https://opensource.org/licenses/Artistic-2.0
+
+'use strict';
+
+const _ = require('lodash');
+const async = require('async');
+const docker = new require('dockerode')();
+const http = require('http');
+const ms = require('ms');
+const fs = require('fs-extra');
```
Comment (rmg) [inline acceptance/repository-cloudant/docker.setup.js]: `fs-extra` isn't listed in `package.json`.. but it also doesn't appear to really be fully utilized here since the only method used is in `fs` from core.

File: acceptance/repository-cloudant/package.json [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+    "@loopback/eslint-config": "^4.1.2",
+    "@loopback/repository": "^1.15.2",
+    "@loopback/repository-tests": "^0.6.1",
+    "@types/node": "^10.17.5",
+    "loopback-connector-cloudant": "2.3.2",
+    "typescript": "~3.7.2",
+    "async": "^3.1.0",
+    "debug": "^4.1.1",
+    "lodash": "^4.17.11",
+    "dockerode": "^2.4.3"
```
Comment (rmg) [inline acceptance/repository-cloudant/package.json]: `ms` is missing from this list - it is used directly by `docker.setup.js`

Comment (agnes512) [reply acceptance/repository-cloudant/package.json]: yeah I forgot.. but why is it still working? 0.0

File: acceptance/repository-cloudant/docker.setup.js (hunk 17 above)
Comment (jannyHou) [reply acceptance/repository-cloudant/docker.setup.js]: true `fs` is enough

File: docker.setup.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (jannyHou) [reply docker.setup.js]: @bajtos Thank you for the technical feedback for the setup script, it's actually added in https://github.com/strongloop/loopback-connector-cloudant/blob/master/docker.setup.js, @agnes512 put it in her PR before that script merged into the master.

To be honest to save the time, the code in that script is mostly copied from an existing script that has been running for test from 2 years ago 🙈 , so for some of your questions I don't quite know the answer, but I'll try :p

File: docker.setup.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (jannyHou) [reply docker.setup.js]: @bajtos I understand that an top-level async function returns a promise which needs to be handled as https://stackoverflow.com/questions/46515764/how-can-i-use-async-await-at-the-top-level, but `async.waterfall()` doesn't return a promise, it just executes functions in sequence then returns `undefined` IIUC...

File: package.json (hunk 11 above)
Comment (jannyHou) [reply package.json]: it will be loaded from dependency `loopback-connector-cloudant`.

File: acceptance/repository-cloudant/package.json (hunk 18 above)
Comment (rmg) [reply acceptance/repository-cloudant/package.json]: `ms` is a pretty common dependency. If any of the other dependencies depend on it then they would have pulled it in and due to dependency tree flattening, it would be installed as a top-level dependency.

File: package.json (hunk 11 above)
Comment (agnes512) [reply package.json]: @jannyHou I moved the whole script to the acceptance/repo-cloudant. But I am okay with either way~

File: acceptance/repository-cloudant/README.md (hunk 14 above)
Comment (agnes512) [reply acceptance/repository-cloudant/README.md]: Done, and README is updated. ( the script is run in this folder now)

File: docker.setup.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (bajtos) [reply docker.setup.js]: > Thank you for the technical feedback of the setup script, it's actually added in https://github.com/strongloop/loopback-connector-cloudant/blob/master/docker.setup.js, @agnes512 put it in her PR before that script merged into the master.

I see, makes sense.

> To be honest, to save the time, the code in that script is mostly copied from an existing script that has been running for test from 2 years ago

Sure, no need to refactor the script if it's working well as it is now. I am going to mark all my comments for `docker.setup.js` as resolved then.

File: acceptance/repository-cloudant/README.md [hunk 19]
```diff
@@ -45,27 +39,25 @@ same (except couchDB doesn't support geosearch).
...
-
-- Start the container, and create an admin and password for this image at
-  `http://localhost:{YOURPORT}/_utils/` if it doesn't have any yet.
-
-- Then run the following script
+- Run the following script
 
 ```bash
-source setup.sh <HOST> <PORT> <USER> <PASSWORD> <DATABASE>
+CLOUDANT_USER=<USER> CLOUDANT_PASSWORD=<PASSWORD> CLOUDANT_DATABASE=<DATABASE> node docker.setup.sh
```
Comment (bajtos) [inline acceptance/repository-cloudant/README.md]: `node docker.setup.sh` looks suspicious - node typically expects `.js` files, not `.sh`. I think it's just a typo?

```suggestion
CLOUDANT_USER=<USER> CLOUDANT_PASSWORD=<PASSWORD> CLOUDANT_DATABASE=<DATABASE> node docker.setup.js
```

BTW why `node docker.setup.js` and not `npm run docker:setup`? Should we modify Travis CI config to call `node docker.setup.js` too?

File: packages/repository-tests/src/crud/replace-by-id.suite.ts (hunk 3 above)
Comment (bajtos) [reply packages/repository-tests/src/crud/replace-by-id.suite.ts]: I see that you added `revisionTokenRequired` flag to fix test failures.

Personally, I'd prefer to find a way how to define `_rev` property only when running on Cloudant, so that we are not using `_rev` property when testing e.g. MongoDB.

Having said that, I don't have any solution I could offer from the top of my head, so it's probably better to accept the current proposal for now and get this pull request landed sooner rather than later 👍

File: packages/repository-tests/src/types.repository-tests.ts [hunk 20]
```diff
@@ -64,6 +64,14 @@ export interface CrudFeatures {
...
    */
   supportsInclusionResolvers: boolean;
+
+  /**
+   * Does the connector require the revision token for replacement operations?
+   * (Cloudant)
+   *
+   * Default: `false`
+   */
+  revisionTokenRequired: boolean;
```
Comment (bajtos) [inline packages/repository-tests/src/types.repository-tests.ts]: Nitpick - feel free to ignore.

The `_rev` value is not always required, is it? For example, when creating a new model instance, the `_rev` must not be provided, at least AFAIK.

Maybe `hasRevisionToken` would be a better name?

File: acceptance/repository-cloudant/README.md (hunk 19 above)
Comment (agnes512) [reply acceptance/repository-cloudant/README.md]: Good catch!

And right, I think 
```
CLOUDANT_USERNAME=<USER> CLOUDANT_PASSWORD=<PASSWORD> CLOUDANT_DATABASE=<DATABASE> npm run docker:setup
```
 also allows users to have their own configurations 👍 changed.

File: packages/repository-tests/src/types.repository-tests.ts (hunk 20 above)
Comment (agnes512) [reply packages/repository-tests/src/types.repository-tests.ts]: make sense

File: acceptance/repository-cloudant/package.json [hunk 21]
```diff
@@ -0,0 +1,49 @@
+{
+  "name": "@loopback/test-repository-cloudant",
+  "version": "1.1.1",
```
Comment (nabdelgadir) [inline acceptance/repository-cloudant/package.json]: I think it should be `1.0.0`

File: acceptance/repository-cloudant/src/__tests__/cloudant-default-repository.acceptance.ts [hunk 22]
```diff
@@ -0,0 +1,20 @@
...
+  CrudRepositoryCtor,
+  crudRepositoryTestSuite,
+} from '@loopback/repository-tests';
+import {CLOUDANT_CONFIG, CLOUDANT_FEATURES} from './cloudant.datasource';
+
+describe('CLOUDANT + DefaultCrudRepository', () => {
+  crudRepositoryTestSuite(
+    CLOUDANT_CONFIG,
+    // Workaround for https://github.com/microsoft/TypeScript/issues/31840
+    DefaultTransactionalRepository as CrudRepositoryCtor,
```
Comment (nabdelgadir) [inline acceptance/repository-cloudant/src/__tests__/cloudant-default-repository.acceptance.ts]: Since it [doesn't support transactions](https://github.com/strongloop/loopback-next/pull/3968/files#diff-b64c9eeda3125036291ac7b928f708a6R22), I think you should change this to `DefaultCrudRepository` like in the MongoDB one.

File: packages/repository-tests/src/crud/replace-by-id.suite.ts (hunk 3 above)
Comment (nabdelgadir) [inline packages/repository-tests/src/crud/replace-by-id.suite.ts]: typo: `cloundant` -> `cloudant`

File: acceptance/repository-cloudant/package.json (hunk 21 above)
Comment (jannyHou) [reply acceptance/repository-cloudant/package.json]: +1

File: acceptance/repository-cloudant/src/__tests__/cloudant-default-repository.acceptance.ts (hunk 22 above)
Comment (jannyHou) [reply acceptance/repository-cloudant/src/__tests__/cloudant-default-repository.acceptance.ts]: 👍 Cloudant uses `_rev` to handle the update conflict, it doesn't handle transaction like sql dbs.

Comment (agnes512) [reply acceptance/repository-cloudant/src/__tests__/cloudant-default-repository.acceptance.ts]: right, I just copied it from MySQL :p
